### PR TITLE
unittest runner: avoid tearDown and cleanup to ease post mortem debugging

### PR DESCRIPTION
**Pull request by mbyt**
Assuming pytest is used to run a GUI test unittest suite. Typically GUI tests close the GUI in the unittest tearDowns or cleanup methods.

These tests cannot be easily post mortem debugged (`pytest --pdb`), as unittests tearDown and cleanup (which close the GUI) are called on an exception before the post mortem debugger jumps into place. Therefore, the GUI is closed and no direct interaction is possible any more.

A work around when running the unittest via the unittest runner is to add the following line to ~/.pdbrc: `!import unittest; unittest.TestCase.run = lambda self,*args,**kw: unittest.TestCase.debug(self)`, start the tests with `python -m pdb test.py` and hit continue.

This pull request accomplishes the same within the pytest unittest compatibility layer. It is solely activated if pytest ist called with --pdb.

An example program which shows the problem would be the following (run with and without this change):
```
from __future__ import print_function
import unittest

class Blub(unittest.TestCase):
    def setUp(self):
        self.addCleanup(lambda: print("2 Cleanup"))

    def tearDown(self):
        print("1 tearDown")

    @classmethod
    def tearDownClass(self):
        print("3 tearDownClass")

    def test_false(self):
        self.assertTrue(False)

def suite():
    return unittest.makeSuite(Blub, 'test_')

if __name__ == '__main__':
    unittest.main(defaultTest='suite')
```

### CODE REVIEW

**Comment by mbyt** [comment]
Please let me know what you think about the problem. Is there another / more simple solution to this problem?

In case the propose solution is sensible, I'll add a corresponding test.

**Comment by nicoddemus** [comment]
Hey @mbyt, thanks for the PR.

TBH I'm reluctant of adding such an esoteric use-case to pytest core, specially on the `unittest` layer. It seems problematic to simply disable teardowns completely when pytest is invoked with `--pdb`. On the other hand, one usually doesn't use the `--pdb` option unless it is running a specific test which is failing and he/she wants to debug.

You can accomplish something similar by adding this to your `conftest.py`:

```python
def pytest_configure(config):
    if config.getvalue("usepdb"):
         import _pytest.unittest
         _pytest.unittest.TestCaseFunction.runtest = lambda self,*args,**kw: self._testcase.debug()
```

**Comment by nicoddemus** [comment]
Hmm now that I thought a little more and played around a bit, I think actually we somehow should make it possible to use `--pdb` with `unittest` subclasses. 

Consider:

```python
import unittest

class Test(unittest.TestCase):

    def tearDown(self):
        self.filename = None

    def test_1(self):
        assert 1

    def test_2(self):
        self.filename = 'debug me'
        assert 0
```

If `--pdb` is used, without @mbyt's PR there's no way to evaluate in the debugger what `self.filename` was on the failing `test_2`, so it is very limited at that point for actually debugging. 

One might even argue that `--pdb` is currently broken when used in `unittest` sub-classes.

What do others think?

**Comment by RonnyPfannschmidt** [comment]
i believe this is something we want to add, but it should be accompanied with a test so we dont break it

**Comment by mbyt** [comment]
Thank you very much for your insights and your fast reply.

I added a corresponding test.

Another possible solution would be to read the `.pdbrc` during the initialization, which would allow to enable monkey patches as the one described.

File: AUTHORS [hunk 1]
```diff
@@ -91,6 +91,7 @@ Martin Prusse
 Matt Bachmann
 Matt Williams
 Matthias Hafner
+mbyt
```
Comment (nicoddemus) [inline AUTHORS]: Don't you want to use your real name here?

File: CHANGELOG.rst [hunk 2]
```diff
@@ -18,15 +18,22 @@
   if a test suite uses ``pytest_plugins`` to load internal plugins (`#1888`_).
   Thanks `@jaraco`_ for the report and `@nicoddemus`_ for the PR (`#1891`_).
 
+* Do not call tearDown (and cleanups) when running unittest with ``--pdb``
```
Comment (nicoddemus) [inline CHANGELOG.rst]: I think we should make crystal clear that this change only affects unittest-tests, so I suggest replacing:

> Do not call tearDown (and cleanups) when running unittest with ``--pdb``

With:
 
> Do not call tearDown and cleanups when running tests from `unittest.TestCase` subclasses with ``--pdb``

File: _pytest/unittest.py [hunk 3]
```diff
@@ -150,7 +150,12 @@ def stopTest(self, testcase):
         pass
 
     def runtest(self):
-        self._testcase(result=self)
+        if self.config.pluginmanager.get_plugin("pdbinvoke") is None:
+            self._testcase(result=self)
+        else:
+            # disables tearDown and cleanups for post mortem debugging
```
Comment (nicoddemus) [inline _pytest/unittest.py]: Please add "(see #1890)" to this comment for future reference

**Comment by nicoddemus** [comment]
Thanks! :grin: 

Other than my comments above, would you mind adding a small blurb to the docs mentioning this change? I think adding a `.. note::` to [unittest.rst](https://github.com/pytest-dev/pytest/blob/master/doc/en/unittest.rst) right before the [Mixing pytest fixtures into unittest.TestCase style tests](https://github.com/pytest-dev/pytest/blob/master/doc/en/unittest.rst#mixing-pytest-fixtures-into-unittesttestcase-style-tests) section would be great, because the previous paragraph mentions using `--pdb` for debugging.

File: AUTHORS (hunk 1 above)
Comment (mbyt) [inline AUTHORS]: Thanks for letting me know. If you do not mind, let's leave it like this.

**Comment by mbyt** [comment]
@nicoddemus thanks again for looking at the changesets and your constructive comments. I modified the code accordingly.

**Comment by nicoddemus** [comment]
Thanks again!

**Comment by The-Compiler** [comment]
I'm late to the party, but this should probably have gone to `features`, no? :wink:

**Comment by nicoddemus** [comment]
Not sure, it seemed to me a bug because you couldn't really use `--pdb` with `TestCase` subclasses.

**Comment by blueyed** [comment]
Just stumbled across this via https://github.com/pytest-dev/pytest/issues/1932.
See https://github.com/pytest-dev/pytest/issues/1932#issuecomment-482886008 for my comment on it.



> These tests cannot be easily post mortem debugged (`pytest --pdb`), as unittests tearDown and cleanup (which close the GUI) are called on an exception before the post mortem debugger jumps into place. Therefore, the GUI is closed and no direct interaction is possible any more.

Why is pdb not invoked before the teardown, if the error is before it?
I think that's the real issue, and just skipping setup/teardown for all tests in general, when `--pdb` is used (which some might even use as a default..!) is rather bad.

As for such tests in general: wouldn't/shouldn't setUp also create the GUI in the first place, which is then skipped?!

**Comment by blueyed** [comment]
Ok, current code looks a bit differrent:
https://github.com/blueyed/pytest/blob/1beba4e69517e69c872fc4a04bed81f9bae1610a/src/_pytest/unittest.py#L219-L226

It might be fine then - let's followup on #1932 / close it if fixed.

**Comment by blueyed** [comment]
https://github.com/pytest-dev/pytest/pull/6014 will revert this.

I think to cover the initial use case a new option like `--unittest-debug` should be added (assuming that we cannot fix/support this properly (https://github.com/pytest-dev/pytest/pull/5996).

**Pull request merged**